extensions/default/src/validateInstances.js

```diff
@@ -0,0 +1,228 @@
+import { vec3 } from 'gl-matrix';
+import toNumber from '@ohif/core/src/utils/toNumber';
+import sortInstancesByPosition from '@ohif/core/src/utils/sortInstancesByPosition';
+import {
+  hasPixelMeasurements,
+  hasOrientation,
+  hasPosition,
+  _getPerpendicularDistance,
+  _getSpacingIssue,
+  reconstructionIssues,
+  _isSameOrientation,
+  constructableModalities,
+} from '@ohif/core/src/utils/isDisplaySetReconstructable';
+
+function addWarning(warnings, warning) {
```

Comment:
I agree that this function shouldnt be here, but where is the ebst place to put it? @ohif/core/src/utils? Right now warnings is a list of strings. It could be renamed to issues, that could be warnings or errors, and it could be a list of objects

Reply:
So I do not know what the requirements are. But if only one extension (like you currently have) is creating the warnings and adding them to each display set, then the process you have defined for adding warnings is fine.

Adding a code for each warning so that it can be translated is a good idea too. The warning/message code is not only useful for translation, but also for maybe in the future for a button/tool to decide if it should be disabled or not based on the code. In my opinion, instead of the message deciding what it applies to, the UI/area decides based on the code how it interprets the message code.

If indeed it is a requirement for any extension/anywhere to add a warning then right now, the only way I can think of allowing extensions (and other places) to add warnings is maybe via the `DisplaySetService` (e.g. maybe `DisplaySetService.addDisplaySetMessage`). The reasons I say `DisplaySetService` are:
- a message/warning is tied to a display set
- if anywhere in the code (e.g. an extension) can add a warning then a notification (e.g. via `DisplaySetService._broadcastEvent(EVENTS.DISPLAY_SET_MESSAGE_ADDED)`) should be sent to the listeners (e.g. `PanelStudyBrowser`) that might care about those messages and react accordingly

Again I am not the expert and I might be totally wrong. These are my humble suggestions. 😊

Reply:
I created two objects in the ohif core: displaySetMessage object that has three attributes: id (number, represents which message), displayLocation (number, where to put the message), hangingProtocolId (string, name of the protocol id to be used in the future). I also created displaySetMessageList, that handles a list of displaySetMessages. 

I think these two objects are sufficient for now. I prefer not to commit to creating a method in displaySetService for now until we better understand what we want to do.

Reply:
Thanks @rodrigobasilio2022.

I will not have time to further review the code until Tuesday.

However, I still believe that a `Message` object should NOT have a `displayLocation` and should NOT have a `hangingProtocolId`. Extensions, modes and UIs should decide themselves which messages are applicable and where to show the messages. Take the following use case...

Use Case A
A hospital adds a their own OHIF mode and wants to display a certain message (id) in a side panel. In the proposed design, would they have to alter the existing/default extension where the message is created and the display location is set to change the location? That feels awkward to me. Sites using OHIF would typically not want to edit the core library code or default extensions to make their own customizations.

Likewise goes for the case where a hospital adds its own extension and hanging protocol. Would they have to edit the message in the default extension or OHIF core to apply or disallow the hanging protocol?

I might be wrong here, but we should think about this.

Reply:
But where to display and which hangingProtocolId to be used, if any,  is a decision of the extension. Right now the extension can only create messages for thumbnails, with no hangingprotocolIds but that is a first version because right now we don't have any demands for location and hanging protocol just for messages in thumbnails. I can remove the fields too

Reply:
Removed both attirbutes.